plugins/crate-copy-azure/src/main/java/io/crate/copy/azure/AzureFileInput.java

```diff
@@ -0,0 +1,164 @@
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.opendal.AsyncOperator;
+import org.apache.opendal.Entry;
+import org.apache.opendal.Operator;
+import org.elasticsearch.common.settings.Settings;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.VisibleForTesting;
+
+import io.crate.execution.engine.collect.files.FileInput;
+import io.crate.execution.engine.collect.files.Globs.GlobPredicate;
+
+/**
+ * File reading components operate with URI.
+ * All URI-s in the public API follow the contract "outgoing/incoming" URI is Azure compatible.
+ * This is accomplished by transforming user provided URI to the Azure compatible format only once.
+ * Outgoing URI-s are then used by other components and sent back to this component,
+ * so outgoing format (expandURI) implicitly dictates incoming URI-s format (getStream).
+ */
+public class AzureFileInput implements FileInput {
+
+    private static final Pattern HAS_GLOBS_PATTERN = Pattern.compile("^((azblob://)[^\\*]*/)[^\\*]*\\*.*");
+
+    private final Map<String, String> config;
+    private final URI uri;
+    private final GlobPredicate uriPredicate;
+    private final String preGlobPath;
+    private final SharedAsyncExecutor sharedAsyncExecutor;
+
+    static class WrapperInputStream extends InputStream {
+
+        private final InputStream delegate;
+        private final Operator operator;
+
+        public WrapperInputStream(InputStream delegate, Operator operator) {
+            this.delegate = delegate;
+            this.operator = operator;
+        }
+
+        @Override
+        public int read() throws IOException {
+            return delegate.read();
+        }
+
+        @Override
+        public void close() throws IOException {
+            delegate.close();
+            operator.close();
+        }
+    }
+
+    /**
+     * @param uri is in user provided format (azblob://path/to/dir)
+     */
+    public AzureFileInput(SharedAsyncExecutor sharedAsyncExecutor, URI uri, Settings settings) {
+        this.sharedAsyncExecutor = sharedAsyncExecutor;
+        config = AzureBlobStorageSettings.openDALConfig(settings);
+        // Pre-glob path operates with user-provided URI as GLOB pattern reflects user facing COPY FROM uri format.
+        this.preGlobPath = toPreGlobPath(uri);
+
+        String resourcePath = resourcePath(uri);
+        this.uri = URI.create(resourcePath);
+        // Glob predicate operates with normalized resource URI to reflect OpenDAL List API response format.
+        // List API returns entries without leading backslash.
+        this.uriPredicate = new GlobPredicate(resourcePath.substring(1));
+    }
+
+    /**
+     * @return List<URI> in Azure compatible format.
+     */
+    @Override
+    public List<URI> expandUri() throws IOException {
+        try (Operator operator = operator()) {
```

Comment:
Updated. Please note, that initially I wanted to close `FileInput` on `closeReader` but those 2 have different lifecycles:

We can have reader/stream closed but `currentInput` still open:
https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java#L326-L327

Hence, closing `FileInput` in `reset` before we nullify it. 
I don't catch exceptions on `currentInput.close()`: my thinking is that we should propagate `reset` exceptions further?

Added no-op default implementation of `close` to the `FileInput` as other implementations don't have anything to close.

Unit testing was a bit cumbersome and all this boilerplate with having `operator()` method and calling instead of using `this.operator` in `expandURI` is caused by not wanting to create `Operator` outside of `AzureFileInput`.

Reply:
Pushed another commit to make `FileOutput` `AutoCloseable`.

Simply added new field `FileOutput` to the `RowWriter`. 

Tried to also remove `OutputStream` field from the `RowWriter` and acquire it based on `FileOutput` but passing `acquireOutputStream` params (or Function to apply) looked more like introducing more mess so I just added a new field.